shell/src/chain_manager.rs

```diff
@@ -324,36 +327,38 @@ impl ChainManager {
         match msg {
             NetworkChannelMsg::PeerBootstrapped(peer_id, peer_metadata, _) => {
                 let peer =
-                    PeerState::new(peer_id, &peer_metadata, chain_state.data_queues_limits());
+                    PeerState::new(peer_id.clone(), &peer_metadata, chain_state.data_queues_limits());
                 // store peer
-                let actor_uri = peer.peer_id.peer_ref.uri().clone();
-                self.peers.insert(actor_uri.clone(), peer);
+                self.peers.insert(peer_id.address, peer);
                 // retrieve mutable reference and use it as `tell_peer()` parameter
-                if let Some(peer) = self.peers.get_mut(&actor_uri) {
+                if let Some(peer) = self.peers.get_mut(&peer_id.address) {
                     tell_peer(
+                        network_channel.clone(),
+                        peer,
                         GetCurrentBranchMessage::new(chain_state.get_chain_id().as_ref().clone())
                             .into(),
-                        peer,
                     );
                 }
             }
-            NetworkChannelMsg::PeerStalled(actor_uri) => {
+            NetworkChannelMsg::PeerDisconnected(peer)
+            | NetworkChannelMsg::PeerBlacklisted(peer) => {
                 // remove peer from inner state
-                if let Some(mut peer_state) = self.peers.remove(&actor_uri) {
+                if let Some(mut peer_state) = self.peers.remove(&peer) {
+                    let peer_id = peer_state.peer_id.clone();
                     // clear innner state (not needed, it will be drop)
                     peer_state.clear();
+                    if let Some(peer_branch_bootstrapper) = self.chain_state.peer_branch_bootstrapper()
+                    {
+                        peer_branch_bootstrapper.tell(CleanPeerData(peer_id), None);
+                    }
                 }
                 // tell bootstrapper to clean potential data
-                if let Some(peer_branch_bootstrapper) = self.chain_state.peer_branch_bootstrapper()
```

Comment:
@binier 
I would better leave this as it was, outside that if,
for example, if `self.peers.remove` was called before in `impl Receive<SystemEvent>` 
then you possibly dont clean peer_branch_bootstrapper

Reply:
I removed that because bootstrapper shouldn't be deciding whom to disconnect. Could cause lots of issues like what if state machine didnt receive event that the peer was disconnected, we might end up bootstrapper having 0 peers and state machine having maxed out connections. Then the bootstrap process would be stuck until someone disconnects and new peer gets connected.

Instead state machine is notified about a stalled peer and it disconnects him and issues an event that the peer was disconnected. Only then we can remove that peer from bootstrapper's state.